src/core/address_payload_unification_design.md

```diff
@@ -0,0 +1,556 @@
+# Design: Address–payload unification
+
+The purpose of this document is to describe a system by which we can
+remove the hard separation between node addresses and payloads, as
+described in [sourcecred/sourcecred#190][issue-190].
+
+[issue-190]: https://github.com/sourcecred/sourcecred/issues/190
+
+This document typechecks—run its source through [`print_code_fences`]
+and copy the result into https://flow.org/try.
+
+[`print_code_fences`]: https://gist.github.com/wchargin/e371d4f3047cf5d89e7ff9bb47224243
+
+This document was written at 2b301f91597a4502612f53a62bdf0367b9249bd4;
+temporal references like “current” and “unchanged” have this reference
+point.
+
+## Core structures
+
+The `Address` type is, for now, unchanged:
+
+```javascript
+type Address = {|
+  +pluginName: string,
+  +type: string,
+  +id: string,
+|};
+```
+
+At the core are two interfaces. First is `NodePayload`:
+
+```javascript
+interface NodePayload {
+  address(): Address;
+}
+```
+
+Plugins are expected to have a variety of subclasses of `NodePayload`,
+one for each logical type. These classes can provide domain-specific
+porcelain. For instance, the GitHub plugin might have:
+
+```javascript
+type GithubNodeType = "PULL_REQUEST" | "PULL_REQUEST_COMMENT";  // e.g.
+class GithubNodePayload<+T: {+url: string}> implements NodePayload {
+  +record: T;
+  +_type: GithubNodeType;
+  constructor(record: T, type: GithubNodeType) {
+    this.record = record;
```

Comment:
I'm curious why the record is a public field, as opposed to having a public method that returns the record. Everything else is accessed functionally, is there a reason to be inconsistent on this point?

Reply:
Makes sense. Weak preference for `_record` then.